### ISSUE TITLE
Destructured Parameters' Default Values Aren't Type Checked

### ISSUE THREAD

**Issue by georgi0u** [google/closure-compiler#2531]
```
/**
 * @param {{x: number}=} param1
 */
function foo({x = "not a number"} = {}) { // passes 
  alert(x);
}

foo(); // passes 

foo({x: "also not a number"}); // doesn't pass
```

It would seem that when you use destructured paramaters with a default value that contradicts the type of that parameter, the compiler doesn't complain when it should.

**Comment by georgi0u** [comment]
I would imagine at best, the function's definition would produce a warning, at at least the call that invokes the function with the mis-typed default parameter.

**Comment by shicks** [comment]
@MatrixFrog would know more about plans for typechecking ES6.  Right now, it happens after transpilation, which transforms the code to

```js
/**
 @param {{x:number}=} param1
 */
function foo(param1) {
  var $jscomp$destructuring$var0 = param1 === undefined ? {} : param1;
  var x = $jscomp$destructuring$var0.x === undefined ? "not a number" : $jscomp$destructuring$var0.x;
  alert(x);
}
foo();
foo({x:"also not a number"});
```

It's not obvious from the output (even with preserve type annotations turned on) whether the $jscomp$destructuring$var0 is typed or not.  If it were given the same type as the parameter then I would expect a type error.  If it's just left to infer the type, then it's clear why there's no error for this.

**Comment by MatrixFrog** [comment]
As I said on the other discussion about this, we should probably put a type annotation on x (and maybe on $jscomp$destructuring$var0 too). If someone wanted to do this as a 20% contribution that would be very appreciated. Otherwise I can probably do it sometime in the next week or so.

**Comment by lauraharker** [comment]
This is fixed now, I think since v20181028.

The compiler warns
```
input0:4: WARNING - default value has wrong type
found   : string
required: number
function foo({x = "not a number"} = {}) { // passes 
              ^^^^^^^^^^^^^^^^^^

```
in the [original example](https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%250A%2520*%2520%2540param%2520%257B%257Bx%253A%2520number%257D%253D%257D%2520param1%250A%2520*%252F%250Afunction%2520foo(%257Bx%2520%253D%2520%2522not%2520a%2520number%2522%257D%2520%253D%2520%257B%257D)%2520%257B%2520%252F%252F%2520passes%250A%2520%2520alert(x)%253B%250A%257D%250A%250Afoo()%253B%2520%252F%252F%2520passes%250A%250Afoo(%257Bx%253A%2520%2522also%2520not%2520a%2520number%2522%257D)%253B%2520%252F%252F%2520doesn't%2520pass%250A).
